WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupEpilogueFragment.java

```diff
@@ -476,23 +504,29 @@ public void onUsernameChanged(OnUsernameChanged event) {
             endProgress();
             showErrorDialog(getString(R.string.signup_epilogue_error_generic));
         } else if (mSignupEpilogueListener != null) {
+            mHasMadeUpdates = true;
             AnalyticsTracker.track(mIsEmailSignup
                     ? AnalyticsTracker.Stat.SIGNUP_EMAIL_EPILOGUE_UPDATE_USERNAME_SUCCEEDED
                     : AnalyticsTracker.Stat.SIGNUP_SOCIAL_EPILOGUE_UPDATE_USERNAME_SUCCEEDED);
-            mSignupEpilogueListener.onContinue();
+            updateAccountOrContinue();
         }
     }
 
     protected boolean changedDisplayName() {
-        return !TextUtils.equals(getArguments().getString(ARG_DISPLAY_NAME), mDisplayName);
+        final String savedDisplayName = mAccount.getAccount().getDisplayName();
+        final String initialDisplayName = getArguments().getString(ARG_DISPLAY_NAME);
+        return !TextUtils.equals(!TextUtils.isEmpty(savedDisplayName)
+                ? savedDisplayName : initialDisplayName, mDisplayName);
```

Comment:
I'm not sure to understand the logic here (and in the `changedUsername` function):
- You check that the stored value in the AccountModel is empty.
- If it's empty you use the initial string generated when the fragment starts and compare with the textfield.

This function will return false if the stored value is empty and the initial generated text is "username" and the display name textfield hasn't been changed. So If I'm not mistaken, in this case, the display name won't be changed and will stay empty. 

Note: I haven't checked how the display name in AccountModel can be empty.

Reply:
> I'm not sure to understand the logic here (and in the `changedUsername` function):
> 
> * You check that the stored value in the AccountModel is empty.
> * If it's empty you use the initial string generated when the fragment starts and compare with the textfield.

That's correct. Since separate updates are now being made, the most assertive way I found to verify that the username/display name actually changed was to check the last stored value rather than relying on the values inside the arguments bundle.

Initially, this check only existed in `changedUsername()` and the reason I assumed it could be empty is because there was a similar check inside [`populateViews()`](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupEpilogueFragment.java#L593-L594). The reason why I later replicated this check to `changedDisplayName()` (even though I wasn't sure it could ever be empty) has to do with the case you just mentioned:

> This function will return false if the stored value is empty and the initial generated text is "username" and the display name textfield hasn't been changed. So If I'm not mistaken, in this case, the display name won't be changed and will stay empty.

First of all, thank you for pointing this out. Yes, I think you're correct. If the stored display name is empty and the user decides to keep the generated one, the generated one won't be saved and the stored one will remain empty.

After investigating this a bit further, I now believe this behavior is rooted in a bug already present in the current implementation, where the generated display name won't be saved if the user doesn't change any value in the epilogue screen. You can verify this by doing the following:
1. Follow the steps 1 to 7 described in the "To test" section of this PR.
2. On the epilogue screen, don't change any of the values but take note of the generated display name.
3. Tap **Continue**.
4. On the next screen, go to the _Me_ tab.
5. Verify that the display name is not the same as the one generated in the epilogue screen (in my case it was the same as the username).

_Note: I realize it may be a little tricky to reproduce this, as I think it may depend on the email you're using to sign up, but I was able to consistently do so, even with the latest version of the app in production (13.7)._

In the proposed implementation, I believe this will only happen in the case you mentioned, where the stored display name is empty, which I'm also not sure if it'll ever be.

I actually stumbled upon this problem while testing some scenarios before submitting this PR but (mistakenly) assumed it was a problem with my own implementation. It only started working as expected after I replicated the check from `changedUsername()` to `changedDisplayName()`, which finally explains why it is there as well.

If we can confirm that the stored display name will never be empty, I think it would be possible to simplify the logic of `changedDisplayName()` so that it simply compares the value of the text field with the stored one. What do you think? Maybe @theck13 can help clarify this as well?

Reply:
Thanks for the investigation!

> If we can confirm that the stored display name will never be empty, I think it would be possible to simplify the logic of changedDisplayName() so that it simply compares the value of the text field with the stored one. What do you think? 

Sounds good to me, I haven't checked yet how the AccountModel is initialized, but I guess it's done during user creation POST call (after the email confirmation).

Reply:
I think the `mAccount.getAccount().getDisplayName()` is always not empty after the account is created before showing the signup epilogue screen.  The `createDisplayNameFromEmail()` method exists to create a slightly more user-friendly display name based on the email address rather than the default display name that matches the username.  In that case, we should _always_ save the display name.  We _could_ optimize it a little and save the display name only if it doesn't match the username, but that might be confusing to future maintainers.

Reply:
Thanks for the input, @theck13!

> I think the `mAccount.getAccount().getDisplayName()` is always not empty after the account is created before showing the signup epilogue screen.

Great, that's my understanding too. I'll proceed with the approach I mentioned earlier then.

Reply:
For future reference, I have also applied the same simplification from `changedDisplayName()` to `changedUsername()` as it didn't seem to ever be empty at that point as well. 

A possible future improvement – beyond the scope of this PR – would be to investigate whether it would be possible to remove similar checks being made inside `populateViews()` and `undoChanges()`.